crates/provider/src/native.rs

```diff
@@ -0,0 +1,307 @@
+use std::{
+    self,
+    error::Error,
+    fmt::Debug,
+    process::{Command, Output},
+};
+
+use async_trait::async_trait;
+use serde::Serialize;
+
+use crate::shared::{
+    constants::{DEFAULT_DATA_DIR, DEFAULT_REMOTE_DIR, LOCALHOST},
+    provider::Provider,
+    types::{NamespaceDef, NamespaceMetadata, RunCommandOptions, RunCommandResponse},
+};
+
+pub trait FileSystem {
+    fn create_dir(&mut self, path: impl Into<String>) -> Result<(), Box<dyn Error>>;
+    fn write(
+        &mut self,
+        path: impl Into<String>,
+        content: impl Into<String>,
+    ) -> Result<(), Box<dyn Error>>;
+}
+
+#[derive(Debug, Serialize)]
+struct FilesystemInMemory {}
+
+#[derive(Debug, Serialize, Clone, PartialEq)]
+struct NativeProvider<T: FileSystem + Debug> {
+    /// Namespace of the client
+    namespace:             String,
+    /// path where configuration relies
+    config_path:           String,
+    // variable that shows if debug is activated
+    debug:                 bool,
+    /// the timeout for the client to exit
+    timeout:               u32,
+    /// command sent to client
+    command:               String,
+    /// temporary directory
+    tmp_dir:               String,
+    pod_monitor_available: bool,
+    local_magic_file_path: String,
+    remote_dir:            String,
+    data_dir:              String,
+    filesystem:            T,
+}
+
+impl<T: FileSystem + Debug> NativeProvider<T> {
+    pub fn new(
+        namespace: impl Into<String>,
+        config_path: impl Into<String>,
+        tmp_dir: impl Into<String>,
+        filesystem: T,
+    ) -> Self {
+        let tmp_dir = tmp_dir.into();
```

Comment:
```suggestion
```

Reply:
This is needed because of the way it is used in the returned `Self`:
```rs
           local_magic_file_path: format!("{}/finished.txt", &tmp_dir), // <---  used here
            remote_dir: format!("{}{}", &tmp_dir, DEFAULT_REMOTE_DIR), // <---  here
            data_dir: format!("{}{}", &tmp_dir, DEFAULT_DATA_DIR), // <---  and here
            command: "bash".into(),
            tmp_dir, // <---  before borrowed here

Reply:
Alright, I didn't see the ```tmp_dir``` was moved (and not borrowed) using the field init shorthand, an alternative solution would be (just for info) to set something like this on the parameter:

```rust
use core::fmt::Display;

tmp_dir: impl Into<String> + Copy + Display
```